File: Coder Army/SOLID/ISP(interfaceSegregationPrinciple)/ISP_violated.cpp
#include <bits/stdc++.h>
using namespace std;

class Shape
{
public:
    virtual double area() = 0;
    virtual double volume() = 0;
};

class Square : public Shape
{
private:
    double side;

public:
    Square(double side) : side(side) {}

    double area() override
    {
        return side * side;
    }

    double volume() override
    {
        throw logic_error("volume not applicable for square");
    }
};

class Rectangle : public Shape
{
private:
    double length, width;

public:
    Rectangle(double length, double width) : length(length), width(width) {}

    double area() override
    {
        return length * width;
    }

    double volume() override
    {
        throw logic_error("volume not applicable for Rectangle");
    }
};

class Cube : public Shape
{
private:
    double side;

public:
    Cube(double side) : side(side) {}

    double area() override
    {
        return 6 * side * side;
    }

    double volume() override
    {
        return side * side * side;
    }
};

int main()
{
    Shape *square = new Square(5);
    Shape *rectangle = new Rectangle(3, 4);
    Shape *cube = new Cube(5);

    cout << " area of Square " << square->area() << endl;
    cout << " area of rectangle " << rectangle->area() << endl;
    cout << " area of cube " << cube->area() << endl;
    cout << " volume of cube " << cube->volume() << endl;
    cout << " volume of rectangle " << rectangle->volume() << endl;
    cout << " volume of Square " << square->volume() << endl;

    delete square;
    delete rectangle;
    delete cube;
    return 0;
}

File: Coder Army/SOLID/ISP(interfaceSegregationPrinciple)/Javascript code/ISP_violated.js
class Shape{
  area(){}
  volume(){}
}

class Square extends Shape{
  constructor(side){
    super();
    this.side=side
  }

  area(){
    return this.side*this.side;
  }
  volume(){
    throw new Error("we don't volume for 2-D shapes")
  }

}
class RectAngle extends Shape{
  constructor(length, breadth){
    super()
    this.length=length,
    this.breadth=breadth
  }

  area(){
    return this.length * this.breadth;
  }
  volume(){
    throw new Error("we don't volume for 2-D shapes")
  }
}

class Cube extends Shape{
  constructor(side){
    super()
    this.side=side
  }

  area(){
    return 6*this.side*this.side;
  }
  volume(){
    return this.side*this.side*this.side
  }
}

const square= new Square(4);
const rectangle= new RectAngle(9, 8)
const cube= new Cube(5);

console.log("Square Area:", square.area());
try {
  console.log("Square Volume:", square.volume());
} catch (e) {
  console.error("Square Volume Error:", e.message);
}

console.log("Rectangle Area:", rectangle.area());
try {
  console.log("Rectangle Volume:", rectangle.volume());
} catch (e) {
  console.error("Rectangle Volume Error:", e.message);
}

console.log("Cube Area:", cube.area());
console.log("Cube Volume:", cube.volume());

